### PR TITLE
Issue-2812 - Allow overridden converse methods to accept messages

### LINKED ISSUE #2812: Allow overridden converse methods to accept emitted messages
**Is your feature request related to a problem? Please describe.**
Passing a message to converse() would make it possible to contribute more complex skills to the project (sometimes, raw utterances are not enough). This changes cannot be simply implementer in an overridden skill converse() since this method is called by the SkillManager.

**Describe the solution you'd like**
In SkillManager.handle_converse_request(), add a function signature check to pass either an overridden converse or the default one.

**Describe alternatives you've considered**
Digging through the message stack directly in skill.converse() but that requires more complex changes.

**Additional context**
This feature is awaiting for a PR review.

### PR DESCRIPTION
**Pull request by NeonAndrii**
## Description
Fixes #2812 
In SkillManager.handle_converse_request(), add a function signature check to pass either an overridden converse or the default one.

## How to test
This change should have no effect on existing Mycroft (or compatible) skills and should be testing on a skill that requires skill.converse to have access to messages passed into SkillManager.handle_converse_request.

## Contributor license agreement signed?
Yes
CLA [ ] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Hello, @NeonAndrii, thank you for helping with the Mycroft project! We welcome everyone
into the community and greatly appreciate your help as we work to build an AI
for Everyone.

To protect yourself, the project, and users of Mycroft technologies we require
a Contributor Licensing Agreement (CLA) before accepting any code
contribution. This agreement makes it crystal clear that along with your
code you are offering a license to use it within the confines of this project.
You retain ownership of the code, this is just a license.

Please visit https://mycroft.ai/cla to initiate this one-time signing. Thank
you!

**Comment by JarbasAl** [comment]
you should also add a `message=None` argument to the converse method in `MycroftSkill`

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2813)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2813/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2813/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2813/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2813/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2813/logs/enclosure.log)

**Comment by NeonDaniel** [comment]
> you should also add a `message=None` argument to the converse method in `MycroftSkill`

I suppose this makes sense if including the message becomes the standard for skills going forward. Either way, skills using a different number of parameters than specified in the `MycroftSkill` class will generate warnings.

**Comment by JarbasAl** [comment]
im starting to think maybe converse should be made to only receive the message 

utterances and lang are in the message, so we would be passing duplicated data

the same check for function signature will still work ensuring compatibility with old skills, but every new skill should start assuming they receive the message object only

File: mycroft/skills/skill_manager.py
Comment (JarbasAl) [inline mycroft/skills/skill_manager.py:451]: as commented below, can this check instead be made

```python
if len(signature(skill_loader.instance.converse).parameters) == 1:
    result = skill_loader.instance.converse(message)
```

and the converse method in MycroftSkill also changed?

Comment (NeonAndrii) [reply mycroft/skills/skill_manager.py:451]: sure, on it

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2813)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2813/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2813/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2813/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2813/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2813/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2813))

**Comment by NeonDaniel** [comment]
> im starting to think maybe converse should be made to only receive the message
> 
> utterances and lang are in the message, so we would be passing duplicated data
> 
> the same check for function signature will still work ensuring compatibility with old skills, but every new skill should start assuming they receive the message object only

No objection here; IMO, converse should have the same data available that an intent match would; I see extracting utterances at this level to be for convenience in writing very simple converse methods (i.e. where you're just looking for a "yes" or a "no" in the utterance

File: mycroft/skills/mycroft_skill/mycroft_skill.py
Comment (forslund) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:354]: I wonder if we should remove the `utterances` and `lang` arguments all together... Since the implementation doesn't support passing all three, having the docs / base class list them all.

This would make handling a message the default implementation while still allowing older skills using the utterances and lang signature using inspect.

This would make the utterance + lang signature deprecated in favor for the use of the message

**Comment by forslund** [comment]
I think this is a good change, I made a comment about maybe deprecating the old way completely since right now it's a bit confusing with three arguments and the third isn't actually used.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2813))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2813))

File: mycroft/skills/mycroft_skill/mycroft_skill.py
Comment (NeonDaniel) [reply mycroft/skills/mycroft_skill/mycroft_skill.py:350]: I think this makes sense if the idea is to move all converse interactions to use messages. Then a warning could be generated in skill_manager if a skill uses the old signature.

**Comment by NeonDaniel** [comment]
Anything holding up this PR or any idea when it might be merged? I have some other skills to submit that rely on this change.

File: mycroft/skills/mycroft_skill/mycroft_skill.py
Comment (JarbasAl) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:337]: this should only receive the message argument

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2813))

**Review by NeonDaniel — APPROVED** [review]

**Comment by NeonDaniel** [comment]
Anything required on this PR for approval/merge?

**Review by forslund — APPROVED** [review]
Works just as described

**Pull request merged**
